recipe-client-addon/lib/ShieldPreferences.jsm

```diff
@@ -0,0 +1,133 @@
+      // Add the opt-out-study checkbox to the Privacy preferences when it is shown.
+      case "advanced-pane-loaded":
+        if (!Services.prefs.getBoolPref("browser.preferences.useOldOrganization", false)) {
+          const checkbox = new this.OptOutStudyCheckbox(subject.document);
+          checkbox.inject();
+        }
+        break;
+      // If the FHR pref changes, set the opt-out-study pref to the value it is changing to.
+      case NS_PREFBRANCH_PREFCHANGE_TOPIC_ID:
+        if (data === FHR_UPLOAD_ENABLED_PREF) {
+          const fhrUploadEnabled = Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF);
+          Services.prefs.setBoolPref(OPT_OUT_STUDIES_ENABLED_PREF, fhrUploadEnabled);
+        }
+        break;
+    }
+  },
+
+  /**
+   * Injects the opt-out-study preference checkbox into about:preferences and
+   * handles events coming from the UI for it.
+   */
+  OptOutStudyCheckbox: class {
+    constructor(document) {
+      this.document = document;
+      this.refs = {};
+
+      const container = this.refs.container = document.createElementNS(XUL_NS, "vbox");
+      container.classList.add("indent");
+
+      const hContainer = this.refs.hContainer = document.createElementNS(XUL_NS, "hbox");
+      hContainer.setAttribute("align", "center");
+      container.appendChild(hContainer);
+
+      const checkbox = this.refs.checkbox = document.createElementNS(XUL_NS, "checkbox");
+      checkbox.setAttribute("id", "optOutStudiesEnabled");
+      checkbox.setAttribute("label", "Allow Firefox to install and run studies");
+      checkbox.setAttribute("preference", OPT_OUT_STUDIES_ENABLED_PREF);
+      checkbox.setAttribute("disabled", !Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF));
+      hContainer.appendChild(checkbox);
+
+      const viewStudies = this.refs.viewStudies = document.createElementNS(XUL_NS, "label");
+      viewStudies.setAttribute("id", "viewShieldStudies");
+      viewStudies.setAttribute("href", "about:studies");
+      viewStudies.setAttribute("useoriginprincipal", true);
+      viewStudies.textContent = "View Firefox Studies";
+      viewStudies.classList.add("learnMore", "text-link");
+      hContainer.appendChild(viewStudies);
+
+      // <prefrence> elements for prefs that we need to monitor while the page is open.
+      const optOutPref = this.refs.optOutPref = document.createElementNS(XUL_NS, "preference");
+      optOutPref.setAttribute("id", OPT_OUT_STUDIES_ENABLED_PREF);
+      optOutPref.setAttribute("name", OPT_OUT_STUDIES_ENABLED_PREF);
+      optOutPref.setAttribute("type", "bool");
+
+      // Weirdly, FHR doesn't have a <preference> element on the page, so we create it.
+      const fhrPref = this.refs.fhrPref = document.createElementNS(XUL_NS, "preference");
+      fhrPref.setAttribute("id", FHR_UPLOAD_ENABLED_PREF);
+      fhrPref.setAttribute("name", FHR_UPLOAD_ENABLED_PREF);
+      fhrPref.setAttribute("type", "bool");
+      fhrPref.addEventListener("change", this);
```

Comment:
I'm just a little worried that this object will live longer than it needs, especially since this.refs holds references to all of these elements. Can you inline the event handler here and change OptOutStudyCheckbox to not use this.refs?